src/sysusers/sysusers.c

```diff
@@ -1550,6 +1553,17 @@ static int parse_line(const char *fname, unsigned line, const char *buffer) {
 
                                 path_kill_slashes(i->uid_path);
                         } else {
+                                _cleanup_free_ char *uid = NULL, *gid = NULL;
+                                if (split_pair(resolved_id, ":", &uid, &gid) == 0) {
+                                        r = parse_gid(gid, &i->gid);
+                                        if (r < 0) {
+                                                log_error("Failed to parse GID: %s", id);
+                                                return -EBADMSG;
+                                        }
+                                        i->gid_set = true;
+                                        i->gid_existing_ok = true;
+                                        resolved_id = uid;
```

Comment:
this will result in a double free like this, as well as leaked memory...

You need to free the old resolved_id first, before you override its pointer. And you need to assign NULL to uid, after you stole its object. Note that you are using the gcc cleanup logic here, i.e. the strings are marked _cleanup_free_ and thus will be destroyed when you leave the scope. And this means the data pointed to initially by resolved_id will be leaked as it is never freed, and the data assigned to it later will be freed twice: once in the inner block as "uid", and once in the outer block again...

Reply:
Thanks for your careful review and sorry for so many errors in so few lines :/ (too much work in languages with automatic memory management recently). I fixed them as you suggested.  I appreciate a careful re-review.

My original version had more duplication but less leaks/double-frees:
```                               
                               _cleanup_free_ char *uid = NULL, *gid = NULL;
                               if (split_pair(resolved_id, ":", &uid, &gid) == 0) {
                                        r = parse_uid(uid, &i->uid);
                                        if (r < 0) {
                                                log_error_errno(r, "Failed to parse UID: '%s': %m", id);
                                                return -EBADMSG;
                                        }
                                        i->uid_set = true;

                                        r = parse_gid(gid, &i->gid);
                                        if (r < 0) {
                                                log_error_errno(r, "Failed to parse GID: '%s': %m", id);
                                                return -EBADMSG;
                                        }
                                        i->gid_set = true;
                                        i->gid_existing_ok = true;
                                } else {
                                        r = parse_uid(resolved_id, &i->uid);
                                        if (r < 0) {
                                                log_error_errno(r, "Failed to parse UID: '%s': %m", id);
                                                return -EBADMSG;
                                        }
                                        i->uid_set = true;
                                }
```
But I guess the current (fixed) code is better as it avoids those duplications.